R/params.R

```diff
@@ -161,11 +161,13 @@ params_get_input <- function(param) {
   input <- param$input
   if (is.null(input)) {
     if (!is.null(param$choices)) {
-      ## radio buttons for a small number of choices, select otherwise.
-      if (length(param$choices) <= 4) {
-        input <- "radio"
-      } else {
+      ## select for a large number of choices and multiple choices.
+      if (isTRUE(param$multiple)) {
+        input <- "select"
+      } else if (length(param$choices) > 4) {
         input <- "select"
+      } else {
+        input <- "radio"
```

Comment:
For the case of `length(choices) > 4` and `multiple = FALSE`, we should use `radio`, right?

```suggestion
      ## select for a large number of choices if multiple is not specified
      if (is.null(param$multiple))
        param$multiple <- length(param$choices) > 4)
      input <- if (param$multiple) "select" else "radio"
```

Reply:
No, I think we want to continue to use `select` when there are many choices. Consider a list naming all of the countries in the world. A set of radio buttons would consume way too much real estate.

Radio buttons are reserved for small numbers of choices when you only want to choose one. Using four items as the tipping point felt about right when we were first working on this. Radio buttons work when you've got a "handful" but not more. It's totally subjective, though, and someone can always override that heuristic by declaring `input: radio` or `input: select` in their document.

Reply:
Also: Your suggestion is setting `param$multiple` when it is not user-specified. That changes the behavior of `select` and turns it from a single-selector to a multi-selector.

Many of the parameter fields become arguments to the Shiny UI controls. In this case, we pass `multiple` to `shiny::selectInput()`.

https://rstudio.github.io/shiny/reference/selectInput.html

Reply:
That last comment probably doesn't apply since the value for `param$multiple` does not escape this function.

Sorry for the repeated replies on this one... Your comment and code might be inconsistent. We never want to use radio for a large number of inputs. I think your code accomplishes that, but your PR comment (not the code comment) says otherwise.

At any rate -- I think the tests capture the scenarios accurately. If you think your rewrite is worth it, fine. I worry that setting `param$multiple` like you have is going to confuse the "future us" because we are giving it a TRUE value even when the parameter does not want to allow multiple choices.